File: collect_and_reveal/android/app/src/main/kotlin/com/example/collect_and_reveal/CollectViews/CollectForm.kt
package com.example.collect_and_reveal.CollectViews

import Skyflow.*
import Skyflow.utils.EventName
import android.content.Context
import android.graphics.Color
import android.view.View
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import com.example.collect_and_reveal.utils.DemoCallback
import io.flutter.Log
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

internal class CollectForm(context: Context, id: Int, flutterEngine: FlutterEngine, client: Client, creationParams: Map<String?, Any?>?) :
        PlatformView {
    private val collectForm = LinearLayout(context);

    override fun getView(): View {
        return collectForm
    }

    private val collectContainer = client.container(ContainerType.COLLECT)

    override fun dispose() {}

    init {
        val CHANNEL = "skyflow-collect/${id}"
        collectForm.orientation = LinearLayout.VERTICAL
        val fields = creationParams?.get("fields") as Map<String, ArrayList<String>>
        for((label, values) in fields) {
            val collectInput = CollectElementInput(
                    table=values[0],
                    column=values[1],
                    label=label,
                    type=getElementType(values[2]),
                    inputStyles = getDefaultStyles()
            )

            val textfield = collectContainer.create(context, collectInput, options = CollectElementOptions(format="mm/yy"))
            collectForm.addView(textfield)
        }

        MethodChannel(flutterEngine.dartExecutor, CHANNEL).setMethodCallHandler{ call, result ->
        if (call.method == "COLLECT") {
            Log.d("MC", "Collect has been called")
            this.collectContainer.collect(DemoCallback(result), CollectOptions())
        } else {
            result.notImplemented()
        }}

        collectForm.setBackgroundColor(Color.rgb(255, 255, 255))

    }

    private fun getElementType(type: String): SkyflowElementType {
        if(type == "CARD_NUMBER") {
            return SkyflowElementType.CARD_NUMBER
        } else if (type.equals("CARDHOLDER_NAME")) {
            return SkyflowElementType.CARDHOLDER_NAME
        } else if(type.equals("CVV")) {
            return SkyflowElementType.CVV
        } else if(type.equals("EXPIRATION_DATE")) {
            return SkyflowElementType.EXPIRATION_DATE
        } else if(type.equals("PIN")) {
            return SkyflowElementType.PIN;
        } else {
            return SkyflowElementType.INPUT_FIELD
        }
    }
}

private fun getDefaultStyles() : Styles {
    val padding = Padding(8, 8, 8, 8)
    val base = Style(
            Color.BLACK
            10f,
            padding,
            1,
            R.font.roboto_light,
            Gravity.START,
            Color.BLUE
    )
    val complete = Style(
            Color.GREEN,
            10f,
            padding,
            1,
            R.font.roboto_light,
            Gravity.END,
            Color.GREEN
    )
    val focus = Style(
            Color.BLACK,
            10f,
            padding,
            1,
            R.font.roboto_light,
            Gravity.START,
            Color.GREEN
    )
    val empty = Style(
            Color.YELLOW,
            10f,
            padding,
            1,
            R.font.roboto_light,
            Gravity.CENTER,
            Color.YELLOW
    )
    val invalid = Style(Color.RED, 10f, padding, 1, R.font.roboto_light, Gravity.START, Color.RED)
    return Styles(base, complete, error, focus, invalid)
}

File: collect_and_reveal/ios/Runner/CollectViews/CollectForm.swift
import Flutter
import Skyflow
import UIKit

class CollectForm: NSObject, FlutterPlatformView {

    private var collectContainer: Container<CollectContainer>
    private var args: Dictionary<String, Dictionary<String, [String]>>?
    private var viewId: Int64
    private var messenger: FlutterBinaryMessenger

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments params: Dictionary<String, Dictionary<String, [String]>>,
        binaryMessenger messenger: FlutterBinaryMessenger?,
        client: Client
    ) {
        self.collectContainer = client.container(type: ContainerType.COLLECT)!
        self.args = params
        self.messenger = messenger!
        self.viewId = viewId


        super.init()
    }
    
    class func getDefaultStyles() -> Styles {
        let padding = UIEdgeInsets(top: 8, left: 8, bottom: 8, right: 8)
        let base = Style(
            borderColor: UIColor.black,
            cornerRadius: 10,
            padding: padding,
            borderWidth: 1,
            textColor: UIColor.blue)
        let complete = Style(
            borderColor: UIColor.green,
            cornerRadius: 10,
            padding: padding,
            borderWidth: 2,
            textColor: UIColor.green)
        let focus = Style(
            borderColor: UIColor.blue,
            cornerRadius: 10,
            padding: padding,
            borderWidth: 2,
            textColor: UIColor.blue)
        let error = Style(
            borderColor: UIColor.red,
            cornerRadius: 10,
            padding: padding,
            borderWidth: 2,
            textColor: UIColor.red)
        return Styles(base: base, complete:complete, focus:focus, invalid: error)
    }

    func getElementType(_ type: String) -> ElementType {
        switch type.uppercased() {
        case "CVV":
            return .CVV
        case "CARDHOLDER_NAME":
            return .CARDHOLDER_NAME
        case "CARD_NUMBER":
            return .CARD_NUMBER
        case "EXPIRATION_DATE":
            return .EXPIRATION_DATE
        case "PIN":
            return .PIN
        default:
            return .CARDHOLDER_NAME
        }
    }

    func view() -> UIView {
        let collectForm = UIView()
        let stackView = UIStackView()
        
        collectForm.addSubview(stackView)
        stackView.translatesAutoresizingMaskIntoConstraints = false
        stackView.axis = .vertical
        
        stackView.topAnchor.constraint(equalTo: collectForm.topAnchor, constant: 10).isActive = true
        stackView.leftAnchor.constraint(equalTo: collectForm.leftAnchor).isActive = true
        stackView.rightAnchor.constraint(equalTo: collectForm.rightAnchor).isActive = true
        
        let styles = CollectForm.getDefaultStyles()

        if let fields = args!["fields"] {
            for (label, field) in fields {
                let label = label
                let tablename = field[0]
                let column = field[1]
                let elementType = getElementType(field[2])

        
                let collectInput = CollectElementInput(table: tablename, column: column, inputStyles: styles, label: label, type: elementType)
                let textField = self.collectContainer.create(input: collectInput)
                textField.backgroundColor = .white
                
                stackView.addArrangedSubview(textField)
            }
        }
        
        
        // set collect method call
        FlutterMethodChannel(name: "skyflow-collect/\(self.viewId)", binaryMessenger: self.messenger)
            .setMethodCallHandler({
                (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
                  if(call.method == "COLLECT") {
                      self.collectContainer.collect(callback: DemoCallback(result))
                  } else {
                    result(FlutterMethodNotImplemented)
                  }
            })
        
        
        return collectForm
    }

}
